### DISCUSSION TITLE
Using environtment variables inide .ini file

### DISCUSSION THREAD

**Discussion by Lepiloff** [sqlalchemy/alembic#1149] [Usage Questions] [answered]
Hi There!

I have 3 files in the root directory `alembic.ini` and `env.py` and `.env`. Inside `.ini` file I try to use environtmen variables

**alembic.ini**

    sqlalchemy.url = postgresql://wplay:wplay@localhost/wplay

**env.py**

    # alembic configuration
    import os
    from alembic import context
    
    
    config = context.config
    section = config.config_ini_section
    config.set_section_option(section, "DB_USER", os.getenv("DB_USER"))
    config.set_section_option(section, "DB_PASS", os.getenv("DB_PASS"))
    config.set_section_option(section, "DB_HOST", os.getenv("DB_HOST"))
    config.set_section_option(section, "DB_NAME", os.getenv("DB_NAME"))

**.env**

    DB_USER='wplay'
    DB_PASS='wplay'
    DB_HOST=db # docker image
    DB_NAME='wplay'

When I changed url to

    sqlalchemy.url = postgresql://${DB_USER}:${DB_PASS}@${DB_HOST}/${DB_NAME}

But when running migrations got

    sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not translate host name "$(DB_HOST)" to address: Name or service not known

What Im doing wrong?

**Comment by zzzeek** [comment]
~~do you mean to use an f string?~~ (edit: this is their config file, not python)

```py
sqlalchemy.url = f"postgresql://{DB_USER}:{DB_PASS}@{DB_HOST}/{DB_NAME}"
```

**Comment by Lepiloff** [comment]
Try this 

`sqlalchemy.url = f'postgresql://${DB_USER}:${DB_PASS}@${DB_HOST}/${DB_NAME}'`

and got

`sqlalchemy.exc.ArgumentError: Could not parse rfc1738 URL from string 'f'postgresql://${DB_USER}:${DB_PASS}@${DB_HOST}/${DB_NAME}'`

**Reply by zzzeek** [reply]
~~omit dollar signs~~  (edit: please disregard)

**Reply by zzzeek** [reply]
oh this is in your config file, sorry, was not clear

**Reply by Lepiloff** [reply]
without the $ 
`sqlalchemy.url = f'postgresql://{DB_USER}:{DB_PASS}@{DB_HOST}/{DB_NAME}'`

same result

Also try this form
`sqlalchemy.url = postgresql://%(DB_USER)s:%(DB_PASS)s@%(DB_HOST)s/%(DB_NAME)s`

And got

`configparser.InterpolationMissingOptionError: Bad value substitution: option 'sqlalchemy.url' in section 'alembic' contains an interpolation key 'db_user' which is not a valid option name. Raw value: 'postgresql://%(DB_USER)s:%(DB_PASS)s@%(DB_HOST)s/%(DB_NAME)s'`

**Comment by zzzeek** [comment]
if you want to set keys in a config that can be re-used, you need to use [config_args](https://alembic.sqlalchemy.org/en/latest/api/config.html#alembic.config.Config.params.config_args) and these are then referenced as ``%(variable_name)s`` (edit: this can't be in env.py, config is already consumed at that point).

```py
config = context.config
config.config_args.update({
"DB_USER": os.getenv("DB_USER"),
"DB_PASS": os.getenv("DB_PASS"),
"DB_HOST": os.getenv("DB_HOST"),
"DB_NAME": os.getenv("DB_NAME")
})
```

config file:

```
sqlalchemy.url = postgresql://%(DB_USER)s:%(DB_PASS)s@%(DB_HOST)s/%(DB_NAME)s
```

**Reply by Lepiloff** [reply]
My goal is to make the `url` parameter dynamically change depending on the value of the virtual environment. By any way

After `config.config_args.update` the result is
`configparser.InterpolationMissingOptionError: Bad value substitution: option 'sqlalchemy.url' in section 'alembic' contains an interpolation key 'db_user' which is not a valid option name. Raw value: 'postgresql://%(DB_USER)s:%(DB_PASS)s@%(DB_HOST)s/%(DB_NAME)s'`

So my current config

**alembic.ini**

```
[alembic]
sqlalchemy.url = postgresql://%(DB_USER)s:%(DB_PASS)s@%(DB_HOST)s/%(DB_NAME)s
```

**env.py**
```

import os
from os import environ
from alembic import context


config = context.config

config.config_args.update({
"DB_USER": os.getenv("DB_USER"),
"DB_PASS": os.getenv("DB_PASS"),
"DB_HOST": os.getenv("DB_HOST"),
"DB_NAME": os.getenv("DB_NAME")
})
```

**Reply by zzzeek** [reply]
config is already consumed by the time you get to env.py so you would need to build a [custom frontend](https://alembic.sqlalchemy.org/en/latest/api/commands.html) to alembic to achieve this (NOTE: this is not the correct way to do this use case overall, see next answer coming up)


```py
import os
from alembic.config import CommandLine, Config

class MyCommandLine(CommandLine):
    def main(self, argv=None):
        options = self.parser.parse_args(argv)
        if not hasattr(options, "cmd"):
            self.parser.error("too few arguments")
        else:

            cfg = Config(
                file_=options.config,
                ini_section=options.name,
                cmd_opts=options,
                config_args={  # type: ignore  # note there's a typing error in alembic
                    "DB_USER": os.getenv("DB_USER", ""),
                    "DB_PASS": os.getenv("DB_PASS", ""),
                    "DB_HOST": os.getenv("DB_HOST", ""),
                    "DB_NAME": os.getenv("DB_NAME", "")

                }
            )
            self.run_cmd(cfg, options)

if __name__ == "__main__":
    MyCommandLine().main()
```

**Reply by zzzeek** [reply, marked as answer]
The correct way to do this pattern is to construct the URL inside of env.py

**this is tested and works**

```py
def run_migrations_online() -> None:

    from sqlalchemy import create_engine
    import re
    import os

    url_tokens = {
        "DB_USER": os.getenv("DB_USER", ""),
        "DB_PASS": os.getenv("DB_PASS", ""),
        "DB_HOST": os.getenv("DB_HOST", ""),
        "DB_NAME": os.getenv("DB_NAME", "")
    }

    url = config.get_main_option("sqlalchemy.url")

    url = re.sub(r"\${(.+?)}", lambda m: url_tokens[m.group(1)], url)

    connectable = create_engine(url)

    with connectable.connect() as connection:
        context.configure(
            connection=connection, target_metadata=target_metadata,
            compare_type=True,
            compare_server_default=True,
        )

        with context.begin_transaction():
            context.run_migrations()

```

config file:

```
sqlalchemy.url = postgresql://${DB_USER}:${DB_PASS}@${DB_HOST}/${DB_NAME}

```

**Reply by jarmos-san** [reply]
@Lepiloff can you mark this question as "answered" so that others (like me, as an example) who stumbled across this thread through a Google Search can find the solution quickly? This little help would be much appreciated.

**Reply by CaselIT** [reply]
Marked

**Reply by cfloressuazo** [reply]
I guess this would be the same if I leave the .ini file as it is by default and then, use a way to construct the URL using environment variables in the env.py file? Like this:

```python
def get_url():
    """Generate a URL from the environment variables.
    """
    return "postgresql://%s:%s@%s:%s/%s" % (
        os.environ["POSTGRES_USER"],
        os.environ["POSTGRES_PASSWORD"],
        os.environ["POSTGRES_HOST"],
        os.environ["POSTGRES_PORT"],
        os.environ["POSTGRES_DB"],
    )

def run_migrations_offline() -> None:
    """Run migrations in 'offline' mode.

    This configures the context with just a URL
    and not an Engine, though an Engine is acceptable
    here as well.  By skipping the Engine creation
    we don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the
    script output.

    """

    url = get_url()

    context.configure(
        url=url,
        target_metadata=target_metadata,
        literal_binds=True,
        dialect_opts={"paramstyle": "named"},
    )

    with context.begin_transaction():
        context.run_migrations()


def run_migrations_online() -> None:
    """Run migrations in 'online' mode.

    In this scenario we need to create an Engine
    and associate a connection with the context.

    """

    url = get_url()

    connectable = create_engine(url)

    with connectable.connect() as connection:
        context.configure(
            connection=connection, target_metadata=target_metadata,
            ompare_type=True,
            compare_server_default=True,
        )

        with context.begin_transaction():
            context.run_migrations()


if context.is_offline_mode():
    run_migrations_offline()
else:
    run_migrations_online()
```

**Reply by CaselIT** [reply]
Sure, you can remove the url from the ini file and obtain it in some other way

**Comment by Lepiloff** [comment]
Thank you for your quick help!

**Comment by nannu8481** [comment]
this is in my alembic.py

sqlalchemy.url = ${db_uri}


this is in my .env 
DATABASE_URI=postgresql://NANNU@localhost/python-flask


this is my env.py file
```py
from logging.config import fileConfig

from sqlalchemy import engine_from_config
from sqlalchemy import pool
import os 
import re
from dotenv import load_dotenv

from alembic import context

# this is the Alembic Config object, which provides
# access to the values within the .ini file in use.
config = context.config
load_dotenv()

db_uri = os.environ.get("DATABASE_URI")

print(f"Database URI: {db_uri}")


# Interpret the config file for Python logging.
# This line sets up loggers basically.
if config.config_file_name is not None:
    fileConfig(config.config_file_name)

# add your model's MetaData object here
# for 'autogenerate' support
# from myapp import mymodel
# target_metadata = mymodel.Base.metadata
target_metadata = None

# other values from the config, defined by the needs of env.py,
# can be acquired:
# my_important_option = config.get_main_option("my_important_option")
# ... etc.


def run_migrations_offline() -> None:
    """Run migrations in 'offline' mode.

    This configures the context with just a URL
    and not an Engine, though an Engine is acceptable
    here as well.  By skipping the Engine creation
    we don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the
    script output.

    """
    url = config.get_main_option("sqlalchemy.url")
    # url = db_uri 
    context.configure(
        url=url,
        target_metadata=target_metadata,
        literal_binds=True,
        dialect_opts={"paramstyle": "named"},
    )

    with context.begin_transaction():
        context.run_migrations()


def run_migrations_online() -> None:
    """Run migrations in 'online' mode.

    In this scenario we need to create an Engine
    and associate a connection with the context.

    """
    connectable = engine_from_config(
        config.get_section(config.config_ini_section, {}),
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    with connectable.connect() as connection:
        context.configure(
            connection=connection, target_metadata=target_metadata
        )

        with context.begin_transaction():
            context.run_migrations()


if context.is_offline_mode():
    run_migrations_offline()
else:
    run_migrations_online()

```
it is not wokring

**Comment by MahirMahbub** [comment]
This solution is not working for the async engine. 
`
    Traceback (most recent call last):
  File "/usr/local/bin/alembic", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/local/lib/python3.12/site-packages/alembic/config.py", line 641, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/usr/local/lib/python3.12/site-packages/alembic/config.py", line 631, in main
    self.run_cmd(cfg, options)
  File "/usr/local/lib/python3.12/site-packages/alembic/config.py", line 608, in run_cmd
    fn(
  File "/usr/local/lib/python3.12/site-packages/alembic/command.py", line 403, in upgrade
    script.run_env()
  File "/usr/local/lib/python3.12/site-packages/alembic/script/base.py", line 583, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/usr/local/lib/python3.12/site-packages/alembic/util/pyfiles.py", line 95, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/alembic/util/pyfiles.py", line 113, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/auth_management/migrations/env.py", line 120, in <module>
    run_migrations_online()
  File "/auth_management/migrations/env.py", line 104, in run_migrations_online
    with connectable.connect() as connection:
         ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 3280, in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 146, in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 3304, in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 449, in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 1263, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 712, in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/impl.py", line 179, in _do_get
    with util.safe_reraise():
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/impl.py", line 177, in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 390, in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 674, in __init__
    self.__connect()
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 900, in __connect
    with util.safe_reraise():
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 896, in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/create.py", line 643, in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 617, in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 937, in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 122, in await_only
    raise exc.MissingGreenlet(
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)
`

**Reply by CaselIT** [reply]
You can't blindly copy a sync env end expect it to work using asyncio.
Modify the async version of the env: https://github.com/sqlalchemy/alembic/blob/main/alembic/templates/async/env.py

**Reply by MahirMahbub** [reply]
I just solved the issue. See the code below. It is working successfully.
```

def run_migrations_offline() -> None:
    """Run migrations in 'offline' mode.

    This configures the context with just a URL
    and not an Engine, though an Engine is acceptable
    here as well.  By skipping the Engine creation
    we don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the
    script output.

    """
    url = config.get_main_option("sqlalchemy.url")
    context.configure(
        url=url,
        target_metadata=target_metadata,
        literal_binds=True,
        dialect_opts={"paramstyle": "named"},
    )

    with context.begin_transaction():
        context.run_migrations()


def do_run_migrations(connection: Connection) -> None:
    context.configure(connection=connection, target_metadata=target_metadata)

    with context.begin_transaction():
        context.run_migrations()

```
